File: src/main/java/no/nav/fo/veilarbregistrering/oppfolging/adapter/OppfolgingClient.java
package no.nav.fo.veilarbregistrering.oppfolging.adapter;

import no.nav.common.health.HealthCheck;
import no.nav.common.health.HealthCheckResult;
import no.nav.common.health.HealthCheckUtils;
import no.nav.common.rest.client.RestClient;
import no.nav.common.rest.client.RestUtils;
import no.nav.common.sts.SystemUserTokenProvider;
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer;
import no.nav.fo.veilarbregistrering.oppfolging.HentOppfolgingStatusException;
import no.nav.fo.veilarbregistrering.registrering.bruker.AktiverBrukerFeil;
import no.nav.fo.veilarbregistrering.registrering.bruker.AktiverBrukerResultat;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.time.Duration;

import static java.time.temporal.ChronoUnit.MILLIS;
import static javax.ws.rs.core.HttpHeaders.AUTHORIZATION;
import static javax.ws.rs.core.HttpHeaders.COOKIE;
import static no.nav.common.utils.UrlUtils.joinPaths;
import static no.nav.fo.veilarbregistrering.config.RequestContext.servletRequest;

public class OppfolgingClient implements HealthCheck {

    private static final Logger LOG = LoggerFactory.getLogger(OppfolgingClient.class);

    private static final int HTTP_READ_TIMEOUT = 120000;

    private final String baseUrl;
    private final OkHttpClient client;

    private final SystemUserTokenProvider systemUserTokenProvider;

    public OppfolgingClient(
            String baseUrl,
            SystemUserTokenProvider systemUserTokenProvider) {
        this.baseUrl = baseUrl;
        this.systemUserTokenProvider = systemUserTokenProvider;
        this.client = RestClient.baseClient().newBuilder().readTimeout(Duration.of(HTTP_READ_TIMEOUT, MILLIS)).build();
    }

    public OppfolgingStatusData hentOppfolgingsstatus(Foedselsnummer fnr) {
        Request request = new Request.Builder()
                .url(HttpUrl.parse(baseUrl).newBuilder()
                        .addPathSegment("oppfolging")
                        .addQueryParameter("fnr", fnr.stringValue())
                        .build())
                .header(COOKIE, servletRequest().getHeader(COOKIE))
                .build();

        try (okhttp3.Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new HentOppfolgingStatusException("Hent oppfølgingstatus feilet med status: " + response.code());
            }
            return RestUtils.parseJsonResponseOrThrow(response, OppfolgingStatusData.class);

        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public AktiverBrukerResultat reaktiverBruker(Foedselsnummer fnr) {
        Request request = buildSystemAuthorizationRequest()
                .url(baseUrl + "/oppfolging/reaktiverbruker")
                .method("POST", RestUtils.toJsonRequestBody(new Fnr(fnr.stringValue())))
                .build();

        try (okhttp3.Response response = client.newCall(request).execute()) {
            return behandleHttpResponse(response, baseUrl + "/oppfolging/reaktiverbruker");
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public AktiverBrukerResultat aktiverBruker(AktiverBrukerData aktiverBrukerData) {
        Request request = buildSystemAuthorizationRequest()
                .url(baseUrl + "/oppfolging/aktiverbruker")
                .method("POST", RestUtils.toJsonRequestBody(aktiverBrukerData))
                .build();

        try (okhttp3.Response response = client.newCall(request).execute()) {
            return behandleHttpResponse(response, baseUrl + "/oppfolging/aktiverbruker");
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    void settOppfolgingSykmeldt(SykmeldtBrukerType sykmeldtBrukerType, Foedselsnummer fnr) {
        Request request = buildSystemAuthorizationRequest()
                .url(HttpUrl.parse(baseUrl).newBuilder()
                        .addPathSegments("oppfolging/aktiverSykmeldt/")
                        .addQueryParameter("fnr", fnr.stringValue())
                        .build())
                .method("POST", RestUtils.toJsonRequestBody(sykmeldtBrukerType))
                .build();

        try (okhttp3.Response response = client.newCall(request).execute()) {
            behandleHttpResponse(response, HttpUrl.parse(baseUrl).newBuilder().addPathSegments("oppfolging/aktiverSykmeldt/")
                    .addQueryParameter("fnr", fnr.stringValue())
                    .build().toString());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private Request.Builder buildSystemAuthorizationRequest() {
        return new Request.Builder()
                .header("SystemAuthorization", this.systemUserTokenProvider.getSystemUserToken())
                .header(AUTHORIZATION, "Bearer " + this.systemUserTokenProvider.getSystemUserToken());
    }

    private AktiverBrukerResultat behandleHttpResponse(okhttp3.Response response, String url) throws IOException {
        int status = response.code();

        if (status == 204) {
            return AktiverBrukerResultat.Companion.ok();
        } else if (status == 403) {
            LOG.warn("Feil ved kall mot: {}, response : {}", url, response);
            AktiverBrukerFeilDto aktiverBrukerFeilDto = parse(response);
            return AktiverBrukerResultat.Companion.feilFrom(mapper(aktiverBrukerFeilDto));
        } else {
            throw new RuntimeException(String.format("Uventet respons (%s) ved kall mot %s", status, url));
        }
    }

    private AktiverBrukerFeil mapper(AktiverBrukerFeilDto aktiverBrukerFeilDto) {
        switch (aktiverBrukerFeilDto.getType()) {
            case BRUKER_ER_DOD_UTVANDRET_ELLER_FORSVUNNET:
                return AktiverBrukerFeil.BRUKER_ER_DOD_UTVANDRET_ELLER_FORSVUNNET;
            case BRUKER_MANGLER_ARBEIDSTILLATELSE:
                return AktiverBrukerFeil.BRUKER_MANGLER_ARBEIDSTILLATELSE;
            case BRUKER_KAN_IKKE_REAKTIVERES:
                return AktiverBrukerFeil.BRUKER_KAN_IKKE_REAKTIVERES;
            case BRUKER_ER_UKJENT:
                return AktiverBrukerFeil.BRUKER_ER_UKJENT;
            default:
                throw new IllegalStateException("Ukjent feil fra Arena: " + aktiverBrukerFeilDto.getType());
        }
    }

    static AktiverBrukerFeilDto parse(okhttp3.Response response) throws IOException {
        return RestUtils.parseJsonResponseOrThrow(response, AktiverBrukerFeilDto.class);
    }

    @Override
    public HealthCheckResult checkHealth() {
        return HealthCheckUtils.pingUrl(joinPaths(baseUrl, "/ping"), client);
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/oppfolging/adapter/OppfolgingClient.kt
package no.nav.fo.veilarbregistrering.oppfolging.adapter

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import io.micrometer.core.instrument.Tag
import no.nav.common.health.HealthCheck
import no.nav.common.health.HealthCheckResult
import no.nav.common.health.HealthCheckUtils
import no.nav.common.utils.UrlUtils
import no.nav.fo.veilarbregistrering.bruker.Foedselsnummer
import no.nav.fo.veilarbregistrering.config.RequestContext.servletRequest
import no.nav.fo.veilarbregistrering.http.Headers
import no.nav.fo.veilarbregistrering.http.Json
import no.nav.fo.veilarbregistrering.http.buildHttpClient
import no.nav.fo.veilarbregistrering.log.logger
import no.nav.fo.veilarbregistrering.log.secureLogger
import no.nav.fo.veilarbregistrering.metrics.Events.*
import no.nav.fo.veilarbregistrering.metrics.MetricsService
import no.nav.fo.veilarbregistrering.metrics.TimedMetric
import no.nav.fo.veilarbregistrering.oauth2.AadOboService
import no.nav.fo.veilarbregistrering.oppfolging.AktiverBrukerException
import no.nav.fo.veilarbregistrering.oppfolging.AktiverBrukerFeil
import no.nav.fo.veilarbregistrering.oppfolging.adapter.veilarbarena.SammensattOppfolgingStatusException
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import java.util.concurrent.TimeUnit
import javax.ws.rs.core.HttpHeaders

open class OppfolgingClient(
    private val objectMapper: ObjectMapper,
    private val metricsService: MetricsService,
    private val baseUrl: String,
    private val aadOboService: AadOboService,
    private val tokenProvider: () -> String,

    ) : TimedMetric(metricsService), HealthCheck {

    open fun reaktiverBruker(fnr: Fnr) {
        val url = "$baseUrl/oppfolging/reaktiverbruker"
        val request = Request.Builder().url(url)
            .headers(Headers.buildHeaders(getServiceAuthorizationHeader()))
            .method("POST", RequestBody.create(Json, objectMapper.writeValueAsString(fnr)))
            .build()
        doTimedCall(REAKTIVER_BRUKER) {
            client.newCall(request).execute().use { reaktiveringResponsMapper(it, fnr) }
        }
    }

    private fun reaktiveringResponsMapper(response: Response, fnr: Fnr) {
        if (response.isSuccessful) return
        else {
            when (response.code()) {
                403 -> {
                    val feil = mapper(objectMapper.readValue(response.body()!!.string()))
                    if (feil == AktiverBrukerFeil.BRUKER_KAN_IKKE_REAKTIVERES_FORENKLET) {
                        secureLogger.info("Bruker med fnr ${fnr.fnr} fikk feil ved reaktivering: $feil")
                    }
                    metricsService.registrer(REAKTIVER_BRUKER_FEIL, feil)
                    throw AktiverBrukerException("Feil ved reaktivering av bruker: ${feil.name}", feil)
                }
                else -> {
                    metricsService.registrer(OPPFOLGING_FEIL, Tag.of("aarsak", response.code().toString()))
                    throw RuntimeException("Feil ved reaktivering av bruker: ${response.code()}, ${response.body()?.string()}")
                }
            }
        }
    }

    open fun aktiverBruker(aktiverBrukerData: AktiverBrukerData) {
        val url = "$baseUrl/oppfolging/aktiverbruker"
        val request = Request.Builder().url(url)
            .headers(Headers.buildHeaders(getServiceAuthorizationHeader()))
            .method("POST", RequestBody.create(Json, objectMapper.writeValueAsString(aktiverBrukerData)))
            .build()
        doTimedCall(AKTIVER_BRUKER) {
            client.newCall(request).execute().use { aktiveringResponsMapper(it) }
        }
    }

    private fun aktiveringResponsMapper(response: Response) {
        if (response.isSuccessful) return
        else {
            when (response.code()) {
                403 -> {
                    val feil = mapper(objectMapper.readValue(response.body()!!.string()))
                    metricsService.registrer(AKTIVER_BRUKER_FEIL, feil)
                    throw AktiverBrukerException("Feil ved aktivering av bruker: ${feil.name}", feil)
                }
                else -> {
                    metricsService.registrer(OPPFOLGING_FEIL, Tag.of("aarsak", response.code().toString()))
                    throw RuntimeException("Feil ved aktivering av bruker: ${response.code()}, ${response.body()?.string()}")
                }
            }
        }
    }

    fun aktiverSykmeldt(sykmeldtBrukerType: SykmeldtBrukerType, fnr: Foedselsnummer) {
        val url = "$baseUrl/oppfolging/aktiverSykmeldt?fnr=${fnr.stringValue()}"
        val request = Request.Builder().url(url)
            .headers(Headers.buildHeaders(getServiceAuthorizationHeader()))
            .method("POST", RequestBody.create(Json, objectMapper.writeValueAsString(sykmeldtBrukerType)))
            .build()
        doTimedCall(OPPFOLGING_SYKMELDT) {
            client.newCall(request).execute().use { aktiveringSykmeldtResponsMapper(it) }
        }
    }


    private fun aktiveringSykmeldtResponsMapper(response: Response) {
        if (response.isSuccessful) return
        else {
            when (response.code()) {
                403 -> {
                    val feil = mapper(objectMapper.readValue(response.body()!!.string()))
                    metricsService.registrer(OPPFOLGING_SYKMELDT_FEIL, feil)
                    throw AktiverBrukerException("Feil ved aktivering av sykmeldt bruker: ${feil.name}", feil)
                }
                else -> {
                    metricsService.registrer(OPPFOLGING_FEIL, Tag.of("aarsak", response.code().toString()))
                    throw RuntimeException("Feil ved aktivering av sykmeldt bruker: ${response.code()}, ${response.body()?.string()}")
                }
            }
        }
    }

    fun erBrukerUnderOppfolging(fodselsnummer: Foedselsnummer): ErUnderOppfolgingDto {
        val url = "$baseUrl/v2/oppfolging?fnr=${fodselsnummer.stringValue()}"
        val request = Request.Builder().url(url)
            .headers(Headers.buildHeaders(getAuthorizationFromCookieOrResolveOboToken()))
            .build()
        return doTimedCall {
            client.newCall(request).execute().use {
                if (it.isSuccessful) {
                    it.body()?.string()?.let { bodyString ->
                        objectMapper.readValue(bodyString, ErUnderOppfolgingDto::class.java)
                    } ?: throw RuntimeException("Unexpected empty body")
                } else throw SammensattOppfolgingStatusException("Feil ved kall til oppfolging-api v2: ${it.code()}")
            }
        }
    }

    private fun getAuthorizationFromCookieOrResolveOboToken(): List<Pair<String, String>> {
        if (aadOboService.erAzureAdToken()) {
            return listOf(("Authorization" to "Bearer ${aadOboService.getAccessToken(oppfolgingApi)}"))
        }

        return listOf(
            servletRequest().getHeader(HttpHeaders.COOKIE)?.let { HttpHeaders.COOKIE to it }
                ?: (HttpHeaders.COOKIE to "selvbetjening-idtoken=${servletRequest().getHeader(HttpHeaders.AUTHORIZATION).removePrefix("Bearer ")}")
        )
    }

    private fun getServiceAuthorizationHeader(): List<Pair<String, String>> =
        listOf(HttpHeaders.AUTHORIZATION to "Bearer ${tokenProvider()}")

    private fun mapper(aktiverBrukerFeilDto: AktiverBrukerFeilDto): AktiverBrukerFeil {
        return when (aktiverBrukerFeilDto.type) {
            ArenaFeilType.BRUKER_ER_DOD_UTVANDRET_ELLER_FORSVUNNET -> AktiverBrukerFeil.BRUKER_ER_DOD_UTVANDRET_ELLER_FORSVUNNET
            ArenaFeilType.BRUKER_MANGLER_ARBEIDSTILLATELSE -> AktiverBrukerFeil.BRUKER_MANGLER_ARBEIDSTILLATELSE
            ArenaFeilType.BRUKER_KAN_IKKE_REAKTIVERES -> AktiverBrukerFeil.BRUKER_KAN_IKKE_REAKTIVERES
            ArenaFeilType.BRUKER_KAN_IKKE_REAKTIVERES_FORENKLET -> AktiverBrukerFeil.BRUKER_KAN_IKKE_REAKTIVERES_FORENKLET
            ArenaFeilType.BRUKER_ER_UKJENT -> AktiverBrukerFeil.BRUKER_ER_UKJENT
            else -> throw IllegalStateException("Ukjent feil fra Arena: $aktiverBrukerFeilDto")
        }
    }

    override fun checkHealth(): HealthCheckResult {
        return HealthCheckUtils.pingUrl(UrlUtils.joinPaths(baseUrl, "/ping"), client)
    }

    override fun value() = "veilarboppfolging"

    companion object {
        val client: OkHttpClient = buildHttpClient {
            readTimeout(120L, TimeUnit.SECONDS)
        }
    }
}

